File: bench-suite/Haskell/STLC/src/Spec.hs
module Spec where

import Etna.Lib
import Data.Maybe (fromJust, isJust)
import Impl

prop_SinglePreserve :: Task Expr
prop_SinglePreserve e =
  isJust mt --> mtypeCheck (pstep e) (fromJust mt)
  where
    mt = getTyp [] e

prop_MultiPreserve :: Task Expr
prop_MultiPreserve e =
  isJust mt --> mtypeCheck (multistep 40 pstep e) (fromJust mt)
  where
    mt = getTyp [] e

mtypeCheck :: Maybe Expr -> Typ -> Bool
mtypeCheck (Just e) t = typeCheck [] e t
mtypeCheck Nothing _ = True -- TODO: add back discards

---

sizeSTLC :: Expr -> Int
sizeSTLC = go
  where
    go :: Expr -> Int
    go (Abs _ e) = 1 + go e
    go (App e1 e2) = 1 + go e1 + go e2
    go _ = 1

File: workloads/OCaml/STLC/lib/spec.ml
open Impl
open QCheck

let mt (e : expr) : typ option = getTyp [] e

let isSome (a : 'a option) : bool =
  match a with Some _ -> true | None -> false

let mtypeCheck (e : expr option) (t : typ) : bool =
  match e with Some e' -> typeCheck [] e' t | None -> true

let prop_SinglePreserve (e : expr) : bool option =
  assume (isSome (mt e));
  mt e >>= fun t' -> Some (mtypeCheck (pstep e) t')

let prop_MultiPreserve (e : expr) : bool option =
  assume (isSome (mt e));
  mt e >>= fun t' -> Some (mtypeCheck (multistep 40 pstep e) t')

let rec sizeSTLC (e : expr) : int =
  match e with
  | Abs (_, e) -> 1 + sizeSTLC e
  | App (e1, e2) -> 1 + sizeSTLC e1 + sizeSTLC e2
  | _ -> 1
